### PR TITLE
[FS] Remove GetMaturedDepositsAsync exception

### PR DESCRIPTION
**Pull request by rowandh**
Modified the return type to `Result` so that we can explicitly indicate the expected failure of a call and return an error message which contains information related to the call. Considered returning `null` but we also need to know the current matured block height at which the request failed.

API still needs to return an error response, otherwise behaviour on the API caller end is different.

Still need to handle exceptions as there are other places they can be thrown.

### CODE REVIEW

File: src/Stratis.Features.FederatedPeg/SourceChain/MaturedBlocksProvider.cs [hunk 1]
```diff
@@ -41,15 +41,17 @@ public MaturedBlocksProvider(ILoggerFactory loggerFactory, IDepositExtractor dep
...
+        public async Task<Result<List<MaturedBlockDepositsModel>>> GetMaturedDepositsAsync(int blockHeight, int maxBlocks)
         {
             ChainedHeader consensusTip = this.consensusManager.Tip;
 
             int matureTipHeight = (consensusTip.Height - (int)this.depositExtractor.MinimumDepositConfirmations);
 
             if (blockHeight > matureTipHeight)
             {
-                throw new InvalidOperationException($"Block height {blockHeight} submitted is not mature enough. Blocks less than a height of {matureTipHeight} can be processed.");
+                // We need to return a Result type here to explicitly indicate failure and the reason for failure.
```
Comment (codingupastorm) [inline src/Stratis.Features.FederatedPeg/SourceChain/MaturedBlocksProvider.cs]: Nice.

File: src/Stratis.Features.FederatedPeg/SourceChain/Result.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+﻿namespace Stratis.Features.FederatedPeg.SourceChain
+{
+    /// <summary>
+    /// A generic result type.
+    /// </summary>
+    /// <typeparam name="T">The type of the value to return if the result was successful.</typeparam>
+    public struct Result<T>
```
Comment (codingupastorm) [inline src/Stratis.Features.FederatedPeg/SourceChain/Result.cs]: Okay this seems like an overengineer now haha, don't we use one of these elsewhere? Or borrow one from some package?

**Comment by quantumagi** [comment]
Very interesting. What are the advantages of doing it this way versus throwing an error?

File: src/Stratis.Features.FederatedPeg/SourceChain/Result.cs (hunk 2 above)
Comment (rowandh) [reply src/Stratis.Features.FederatedPeg/SourceChain/Result.cs]: We do but I didn't want to add another dependency to FS.

**Comment by rowandh** [comment]
@quantumagi given that this is an expected case we don't need to throw an exception. Sounds like it's expected fairly often too, so there is probably some associated performance benefit.

**Pull request merged**